pkg/transforms/mqttsecret.go

```diff
@@ -0,0 +1,270 @@
+		keypemblock:  []byte(secrets[MQTTSecretClientKey]),
+		certpemblock: []byte(secrets[MQTTSecretClientCert]),
+		capemblock:   []byte(secrets[MQTTSecretCACert]),
+	}
+
+	return mqttSecrets, nil
+}
+func (sender *MQTTSecretSender) validateSecrets(secrets mqttSecrets) error {
+	caCertPool := x509.NewCertPool()
+	if sender.mqttConfig.AuthMode == AuthModeUsernamePassword {
+		if secrets.username == "" || secrets.password == "" {
+			return errors.New("AuthModeUsernamePassword selected however username or password was not found at secret path")
+		}
+
+	} else if sender.mqttConfig.AuthMode == AuthModeCert {
+		// need both to make a successful connection
+		if len(secrets.keypemblock) <= 0 || len(secrets.certpemblock) <= 0 {
+			return errors.New("AuthModeCert selected however the key or cert PEM block was not found at secret path")
+		}
+	} else if sender.mqttConfig.AuthMode == AuthModeCA && len(secrets.capemblock) <= 0 {
+		return errors.New("AuthModeCA selected however no PEM Block was found at secret path")
+	}
+
+	if len(secrets.capemblock) > 0 {
+		ok := caCertPool.AppendCertsFromPEM([]byte(secrets.capemblock))
+		if !ok {
+			return errors.New("Error parsing CA Certificate")
+		}
+	}
+	return nil
+}
+func (sender *MQTTSecretSender) configureMQTTClientForAuth(secrets mqttSecrets) error {
+	var cert tls.Certificate
+	var err error
+	caCertPool := x509.NewCertPool()
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: sender.mqttConfig.SkipCertVerify,
+	}
+	switch sender.mqttConfig.AuthMode {
+	case AuthModeUsernamePassword:
+		sender.opts.SetUsername(secrets.username)
+		sender.opts.SetPassword(secrets.password)
+	case AuthModeCert:
+		cert, err = tls.X509KeyPair(secrets.certpemblock, secrets.keypemblock)
+		if err != nil {
+			return err
+		}
+		tlsConfig.Certificates = []tls.Certificate{cert}
+	case AuthModeCA:
+		break
+	case AuthModeNone:
+		return nil
+	default:
+		return nil
+	}
+
+	if len(secrets.capemblock) > 0 {
+		ok := caCertPool.AppendCertsFromPEM(secrets.capemblock)
+		if !ok {
+			return errors.New("Error parsing CA PEM block")
+		}
+		tlsConfig.ClientCAs = caCertPool
+	}
+
+	sender.opts.SetTLSConfig(tlsConfig)
+
+	return nil
+}
+func (sender *MQTTSecretSender) initializeMQTTClient(edgexcontext *appcontext.Context) error {
+	if sender.mqttConfig.AuthMode == "" {
+		sender.mqttConfig.AuthMode = "none"
+		edgexcontext.LoggingClient.Warn("AuthMode not set, defaulting to \"none\"")
+	}
+
+	//get the secrets from the secret provider and populate the struct
+	secrets, err := sender.getSecrets(edgexcontext)
+	if err != nil {
+		return err
+	}
+	//ensure that the authmode selected has the required secret values
+	if secrets != nil {
+		err = sender.validateSecrets(*secrets)
+		if err != nil {
+			return err
+		}
+		// configure the mqtt client with the retrieved secret values
+		err = sender.configureMQTTClientForAuth(*secrets)
+		if err != nil {
+			return err
+		}
+	}
+
+	sender.lastRetrieved = time.Now()
```

Comment:
This is unclear what it is. Might help to change `lastRetrieved` to `secretsLastRetrieved` and should this be right after call to getSecrets() for context?